### PR TITLE
parser: merge nested dynamic attributes

### PR DESCRIPTION
**Pull request by ncfavier**
# Motivation

Fixes https://github.com/NixOS/nix/issues/7115

# Context

https://github.com/NixOS/nix/pull/2089

This doesn't completely get rid of the order sensitivity, because `recursive` is still only inherited from the first attrset, so if the second one has `rec` that will be silently ignored. There should probably be an error if `jAttrs->recursive != ae->recursive`, but I guess that's a breaking change?

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by sternenseemann** [comment]
Can you a regression test to the lang test suite covering this?

> This doesn't completely get rid of the order sensitivity, because `recursive` is still only inherited from the first attrset, so if the second one has `rec` that will be silently ignored. There should probably be an error if `jAttrs->recursive != ae->recursive`, but I guess that's a breaking change?

See https://github.com/NixOS/nix/issues/7111, I think also worth fixing, but we can discuss it separately.

**Review by sternenseemann — APPROVED** [review]

**Comment by Ericson2314** [comment]
OK I see the issue now. It just silently drops. Wow!

**Comment by fricklerhandwerk** [comment]
Triaged in the Nix team meeting:

tl;dr This needs release notes, since it's a breaking change. Otherwise approved.

- @edolstra: the root problem is that we should not allow mixing static and dynamic attribute names
  - @Ericson2314: this was introduced when someone explicitly asked for it. now we're obligated to support it and make it correct
- @thufschmitt: it's a breaking change to account for a pretty obscure corner case. we probably wouldn't have noticed if @sternenseemann didn't dig into the language to reimplement it
- @roberth: this is something that should not be relied on. it also doesn't seem like it's widely used (although it's hard to tell)
- @Ericson2314: we should issue a warning (once). but since it may be a transitional thing, once people are used to it we may not need it any more
- @fricklerhandwerk: agree with the proposal to fix it as a bug. if it's an obscure corner case indeed, it shouldn't break too much existing code. and if it does, it will break "in the right direction"
  - @roberth: worst case if someone is affected they can open an issue about a regression
  - @thufschmitt: just worried some obscure NixOS modules may rely on the current behavior
- @Ericson2314: this needs release notes, since it's a breaking change, even though it fixes what we consider a bug

**Comment by ncfavier** [comment]
Added release notes and rebased.

File: doc/manual/src/release-notes/rl-next.md [hunk 1]
```diff
@@ -6,3 +6,16 @@
...
+  ```nix
+  {
+    nested = { foo = 1; };
+    nested = { ${"ba" + "r"} = 2; };
+  }
+  ```
+
+  This used to silently discard `nested.bar`, but now behaves correctly.
+
+  Note that this feature is [controversial](https://github.com/NixOS/nix/pull/2089#issuecomment-382672405) and should not be relied upon too much as it makes files harder to read.
```
Comment (fricklerhandwerk) [inline doc/manual/src/release-notes/rl-next.md:21]:
````suggestion
  Note that the feature of merging multiple attribute set declarations is of questionable value, as it makes files harder to read.
  It has been around for a long time and is therefore supported for backwards compatibility, but should not be relied upon.
````

I don't think we should link to GitHub issues. Just put the information inline.

File: doc/manual/src/release-notes/rl-next.md [hunk 2]
```diff
@@ -6,3 +6,16 @@
...
+- Nested dynamic attributes are now merged correctly by the parser. For example:
+
+  ```nix
+  {
+    nested = { foo = 1; };
+    nested = { ${"ba" + "r"} = 2; };
+  }
+  ```
+
+  This used to silently discard `nested.bar`, but now behaves correctly.
```
Comment (fricklerhandwerk) [inline doc/manual/src/release-notes/rl-next.md:19]:
````suggestion
  This used to silently discard `nested.bar`, but now behaves as one would expect and evaluates to
  
  ```nix
  { nested = { bar = 2; foo = 1; }; }
  ```
````

File: doc/manual/src/release-notes/rl-next.md (hunk 1 above)
Comment (fricklerhandwerk) [reply doc/manual/src/release-notes/rl-next.md]: Sorry, I edited the comment while thinking about it. Please re-check the suggestion.

**Review by fricklerhandwerk — APPROVED** [review]
Thanks. This is an exemplary PR.

**Comment by ncfavier** [comment]
There are two kinds of merges:

```nix
{
  foo.bar = 1;
  foo.qux = 2;
}
{
  foo = { bar = 1; };
  foo = { qux = 2; };
}
```

This is only about the second one. You're right that this is completely unclear.

**Comment by ncfavier** [comment]
https://github.com/NixOS/nix/pull/8741

Feedback welcome, I'm bad at this
